### DISCUSSION TITLE
Some new starter questions + suggestions (cmake/multiple-plugin setups/automation/dependency simplification)

### DISCUSSION THREAD

**Discussion by etheory** [iPlug2/iPlug2#1104] [General]
Hi @olilarkin! Thanks so much for this amazing project.

I am looking to get started with iPlug2 after having used the VST SDK and worked with various GUI API's and I like what I see.

I am a low-level C/C++/assembly developer in the film industry for real-time 3D rendering by day, and an audio DSP developer and analog synthesizer builder by night, so I'm no newbie to programming, but getting started with iPlug2 I had a few questions I could not find through any of the docs or channels:

1.) Why do you hard-code v142 (vs2019) in all the vs project files (or have any included project files for any platform at all when they can be generated automatically via cmake)? Why not just use CMake to generate whatever native format the user wants? Is there a way to do this I'm not aware of? Are you open to collaborative development in this area? Is anyone else actively doing this I could collaborate with? I think a cmake version of your system would really help make everything more automatable. I am using vs2022 so it felt immediately strange to have to do string-replace just to start a project. This could be streamlined.

2.) You use a sub-module for iPlug2 in the iPlug2OOS repo along with scripts to download dependencies. It's more common in my field to locally build a project and use either environment variables or paths in CMake to define how to find them from this local build. Is this version of development possible? I say this because if I want to create a suite of plugins, having to re-run the same thing and download hundreds of megabytes of dependencies per plugin isn't very efficient.

3.) Is there a standard methodology when using iPlug2 to have one project containing multiple-plugins? If so, are there docs for this process? If not would you be open to collaborating on making this possible?

4.) Using cmake would also make it redundant to have multiple build scripts for each OS/API type etc. You could just have one that then gets configured automatically to "do the right thing". Again, are you open to collaborating on this?

I like what I see, but it seems to me to be too setup for single-plugin builds with all dependencies deposited directly into the project. Since I'd like to instead work on a large repository with multiple plugins all centrally handled by one script, I'd like to try and help other people modify iPlug2 to allow this if it doesn't already.

It also feels like iPlug2OOS could be rewritten to just be a custom CMake command you call from a fresh project rather than a project you have to clone to do the basics. Though I guess then you lose the included code from the example project....

Kind regards,
Luke

**Comment by olilarkin** [comment]
Hello, about to go on holiday so apologies, i'll be brief.

Yes the plan is to support CMake properly soon. Several people are using iPlug2 with CMake. 

At the moment it is on a branch here

https://github.com/iPlug2/iPlug2/tree/cmake

and here

https://github.com/iPlug2/iPlug2OOS/tree/cmake

I would use the iPlug2OOS setup for multi-plugin projects these days.

If you have expertise in CMake scripting, it would be great to get some help with it https://github.com/iPlug2/iPlug2/issues/39, 

Basically the reason it is not merged yet is because I'd like to get it to a state where we have parity with the custom made Xcode projects etc. This is not so trivial especially for iOS AUv3 builds, Web Audio Modules etc.

**Comment by etheory** [comment]
As an FYI I copied over the IPlugInstrument example and created a project entirely using CMake to simplify it, on Windows, and it works well:

**buildpaths.cmake**
```
set(IPLUG2_ROOT D:/Programming/iPlug2)

set(IPLUG2_SDK_FILES
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/baseiids.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/fbuffer.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/fdynlib.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/fobject.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/fstreamer.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/fstring.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/timer.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/source/updatehandler.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/thread/source/fcondition.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/thread/source/flock.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/main/dllmain.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/main/pluginfactory.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstaudioeffect.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstbus.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstcomponent.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstcomponentbase.cpp
  #${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vsteditcontroller.cpp
  #${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstgui_win32_bundle_support.cpp
  #${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstguieditor.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstinitiids.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstnoteexpressiontypes.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstparameters.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstpresetfile.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstrepresentation.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/vstsinglecomponenteffect.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/commoniids.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/memorystream.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/openurl.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/pluginview.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/readfile.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/systemclipboard_win32.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/common/threadchecker_win32.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/public.sdk/source/vst/utility/stringconvert.cpp

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/pluginterfaces/base/conststringtable.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/pluginterfaces/base/coreiids.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/pluginterfaces/base/funknown.cpp
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/pluginterfaces/base/ustring.cpp

  ${IPLUG2_ROOT}/IPlug/IPlugParameter.cpp
  ${IPLUG2_ROOT}/IPlug/IPlugPluginBase.cpp
  ${IPLUG2_ROOT}/IPlug/IPlugAPIBase.cpp
  ${IPLUG2_ROOT}/IPlug/IPlugProcessor.cpp
  ${IPLUG2_ROOT}/IPlug/IPlugTimer.cpp
  ${IPLUG2_ROOT}/IPlug/IPlugPaths.cpp

  ${IPLUG2_ROOT}/IPlug/VST3/IPlugVST3.cpp
  ${IPLUG2_ROOT}/IPlug/VST3/IPlugVST3_ProcessorBase.cpp

  ${IPLUG2_ROOT}/IPlug/Extras/Synth/MidiSynth.cpp
  ${IPLUG2_ROOT}/IPlug/Extras/Synth/VoiceAllocator.cpp

  ${IPLUG2_ROOT}/IGraphics/IGraphics.cpp
  ${IPLUG2_ROOT}/IGraphics/IGraphicsEditorDelegate.cpp
  ${IPLUG2_ROOT}/IGraphics/IControl.cpp

  ${IPLUG2_ROOT}/IGraphics/Controls/IControls.cpp
  ${IPLUG2_ROOT}/IGraphics/Controls/IPopupMenuControl.cpp
  ${IPLUG2_ROOT}/IGraphics/Controls/ITextEntryControl.cpp

  ${IPLUG2_ROOT}/IGraphics/Platforms/IGraphicsWin.cpp
)

set(IPLUG2_INCLUDE_DIRS
  ${IPLUG2_ROOT}/WDL

  ${IPLUG2_ROOT}/IPlug

  ${IPLUG2_ROOT}/IPlug/VST3

  ${IPLUG2_ROOT}/IPlug/Extras
  ${IPLUG2_ROOT}/IPlug/Extras/Synth

  ${IPLUG2_ROOT}/IGraphics
  ${IPLUG2_ROOT}/IGraphics/Controls
  ${IPLUG2_ROOT}/IGraphics/Drawing
  ${IPLUG2_ROOT}/IGraphics/Platforms

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK
  ${IPLUG2_ROOT}/Dependencies/IGraphics/NanoSVG/src
  ${IPLUG2_ROOT}/Dependencies/IGraphics/NanoVG/src
  ${IPLUG2_ROOT}/Dependencies/IGraphics/glad_GL2/include
  ${IPLUG2_ROOT}/Dependencies/IGraphics/glad_GL2/src
  ${IPLUG2_ROOT}/Dependencies/IGraphics/STB

  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/base/thread/include
  ${IPLUG2_ROOT}/Dependencies/IPlug/VST3_SDK/pluginterfaces/vst
)
```

**CMakeLists.txt**
```
cmake_minimum_required(VERSION 3.10)

project(etheoryTestPlugins-vst3)

include(buildpaths.cmake)

set(TARGET_NAME etheoryTestInstrument)

set(CMAKE_VERBOSE_MAKEFILE ON)

set(SOURCE_FILES
  src/resources/main.rc
  src/resources/resource.h

  src/config.h
  src/etheoryTestInstrument_DSP.h
  src/etheoryTestInstrument.h
  src/etheoryTestInstrument.cpp
)

add_library(${TARGET_NAME} SHARED
  ${SOURCE_FILES}
  ${IPLUG2_SDK_FILES}
)

target_include_directories(${TARGET_NAME} PUBLIC
  ${IPLUG2_INCLUDE_DIRS}
)

target_link_libraries(${TARGET_NAME}
  wininet
  comctl32
  Shlwapi
)

#set_target_properties(${TARGET_NAME} PROPERTIES SUFFIX .vst3 LINK_FLAGS /NODEFAULTLIB)
set_target_properties(${TARGET_NAME} PROPERTIES SUFFIX .vst3)

target_compile_features(${TARGET_NAME} PUBLIC cxx_std_17)

# NOMINMAX fix usage of std::max and std::min
target_compile_definitions(${TARGET_NAME} PUBLIC
  VST3_API
  NOMINMAX
  IGRAPHICS_NANOVG
  IGRAPHICS_GL2
  IPLUG_EDITOR
  IPLUG_DSP
)

# 4068 disable pragma warning about unknown pragma 'mark'
# 4996 disable pragma warning about unsafe functions used by WDL like _vsnprintf etc.
target_compile_options(${TARGET_NAME} PUBLIC
  /wd4068
  /wd4996
)
```

But I would like to do a find_package cmake script for IPlug2 to simplify all the above, that would work with a VST3 find_package etc. to allow the user to decide the output target that way.

The nice thing about the above (even though it's crude) is that it "just works" no matter your compiler etc. and it integrates nicely with VSCode.

I hope it's useful to someone.

But happy to work with others on the official efforts too.

**Comment by olilarkin** [comment]
Cmake support is now merged 🥳
